File: comparison/java/Quicksort.java

// https://stackoverflow.com/questions/504103/how-do-i-write-a-correct-micro-benchmark-in-java maybe?

import java.util.Collections;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

class Quicksort {
    public static void main(String[] args) {
        long tenmillion = 10000000;
        int maximum = Integer.MAX_VALUE;
        // custom quicksort
        List<Integer> liste = new Random().ints(tenmillion, 0, maximum)
                .boxed()
                .collect(Collectors.toList());

        // liste.forEach(System.out::print);
        long start = System.nanoTime();
        quicksort(liste, 0, liste.size() - 1);
        long end = System.nanoTime();
        System.out.println((end - start) / 1000000);
        // liste.forEach(System.out::print);

        // std sort
        List<Integer> liste_zwei = new Random().ints(tenmillion, 0, maximum)
                .boxed()
                .collect(Collectors.toList());

        long start_zwei = System.nanoTime();
        Collections.sort(liste_zwei);
        long end_zwei = System.nanoTime();
        System.out.println((end_zwei - start_zwei) / 1000000);
    }

    private static int hoare_partition(List<Integer> list, int start, int end) {
        int pivot = list.get((int) ((start + end) / 2));
        int i = start - 1;
        int j = end + 1;
        while (true) {
            i++;
            while (list.get(i) < pivot) {
                i++;
            }
            j--;
            while (list.get(j) > pivot) {
                j--;
            }
            if (i >= j) {
                return j;
            }
            int temp = list.get(i);
            list.set(i, list.get(j));
            list.set(j, temp);
        }
    }

    private static void quicksort(List<Integer> list, int start, int end) {
        if (start >= 0 && end >= 0 && start < end) {
            int crossing = hoare_partition(list, start, end);
            quicksort(list, start, crossing);
            quicksort(list, crossing + 1, end);
        }
    }

}

File: comparison/javascript/quicksort.js
function hoare_partition(array, start, end) {
    pivot = array[Math.trunc((start + end) / 2)];

    i = start - 1;
    j = end + 1;

    while (true) {
        ++i;
        while (array[i] < pivot) {
            ++i;
        }

        --j;
        while (array[j] > pivot) {
            --j;
        }

        if (i >= j) {
            return j;
        }

        temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
}

function quicksort(array, start, end) {
    if (start >= 0 && end >= 0 && start < end) {
        crossing = hoare_partition(array, start, end);

        quicksort(array, start, crossing);
        quicksort(array, crossing + 1, end);
    }
}

// custom quicksort
let liste = Array(10000000)
    .fill()
    .map(() => Math.floor((4294967295 + 1) * Math.random()));

let start = performance.now();
quicksort(liste, 0, liste.length - 1);
let end = performance.now();

console.log("Quicksort", end - start);

// std sort
let liste_zwei = Array(10000000)
    .fill()
    .map(() => Math.floor((4294967295 + 1) * Math.random()));

let start_zwei = performance.now();
quicksort(liste, 0, liste.length - 1);
let end_zwei = performance.now();

console.log("Standard Sort", end_zwei - start_zwei);
